File: Difficulty: Easy/DFS of Graph/dfs-of-graph.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends


class Solution {
  public:
    void dfs_traversal(vector<vector<int>>& adj,vector<int> &ans,unordered_set<int> &found, int node)
    {
        ans.push_back(node);
        found.insert(node);
        
        for(auto it : adj[node])
        {
            if(found.find(it)==found.end())
            {
                dfs_traversal(adj,ans,found,it);
            }
        }
    }
    vector<int> dfs(vector<vector<int>>& adj) {
        // Code here
        vector<int>ans;
        unordered_set<int>found;
        
        for(int i=0;i<adj.size();i++)
        {
            if(found.find(i)==found.end())
            dfs_traversal(adj,ans,found,i);
        }
        
        return ans;
        
    }
};




//{ Driver Code Starts.

int main() {
    int tc;
    cin >> tc;
    cin.ignore();
    while (tc--) {
        int V;
        cin >> V;
        cin.ignore();
        vector<vector<int>> adj(
            V); // Use vector of vectors instead of array of vectors.

        for (int i = 0; i < V; i++) {
            string input;
            getline(cin, input);
            int num;
            vector<int> node;
            stringstream ss(input);
            while (ss >> num) {
                node.push_back(num);
            }
            adj[i] = node;
        }

        Solution obj;
        vector<int> ans = obj.dfs(adj);
        for (int i = 0; i < ans.size(); i++) {
            cout << ans[i] << " ";
        }
        cout << endl;
        cout << "~" << endl;
    }
    return 0;
}

// } Driver Code Ends

File: Difficulty: Easy/DFS of Graph/dfs-of-graph.java
class Solution {
    void find_dfs(ArrayList<Integer> ans,ArrayList<ArrayList<Integer>> adj,Set<Integer>found, int u)
    {
        found.add(u);
        ans.add(u);
        
        for(int i=0;i<adj.get(u).size();i++)
        {
            int v=adj.get(u).get(i);
            
            if(!found.contains(v))
                find_dfs(ans,adj,found,v);
        }
    }
    public ArrayList<Integer> dfs(ArrayList<ArrayList<Integer>> adj) {
        // code here
        
        ArrayList<Integer> ans=new ArrayList<>();
        
        Set<Integer>found=new HashSet<>();
        
        
        for(int i=0;i<adj.size();i++)
        {
            if(!found.contains(i))
                find_dfs(ans,adj,found,i);
        }
        
        return ans;
    }
}